### ISSUE TITLE
The simulating advantage of Box<T>

### ISSUE THREAD

**Issue by 7dy109** [kaist-cp/cs220#225]
Hello, all. First of all, I really appreciate the contribution of TAs to this course. Thank you for making this course wonderful!

In the rust book, "When you have a large amount of data and you want to transfer ownership but ensure the data won’t be copied when you do so" - [rust book](https://doc.rust-lang.org/book/ch15-01-box.html)
The details in rust book in below.
> transferring ownership of a large amount of data can take a long time because the data is copied around on the stack. To improve performance in this situation, we can store the large amount of data on the heap in a box. Then, only the small amount of pointer data is copied around on the stack, while the data it references stays in one place on the heap. 

But When I run below code the result is below.
```
Using Box<T>
4942
Not Using Box<T>
4742
```
```rust
    use std::time::Instant;
    fn takes_ownership(some : Box<Vec<i32>>) {
        println!("Using Box<T>");
    }
    fn takes_ownership2(some : Vec<i32>) {
        println!("Not Using Box<T>");
    }
    let start = Instant::now();
    let test:Box<Vec<i32>> = Box::new((0..214748364).collect());
    takes_ownership(test);
    println!("{}", start.elapsed().as_millis());
    let start2 = Instant::now();
    let test2:Vec<_> = (0..214748364).collect();
    takes_ownership2(test2);
    println!("{}", start2.elapsed().as_millis());
```

Why time need when moving ownership when using box<T> is not shorter than time need when moving ownership when not using box<T>?
Is my simulation code wrong?

**Comment by Lee-Janggun** [comment]
Data in Vec is already in the heap; hence when you wrap it in a box, you simply have a reference to a reference to the heap.

If you want to bench this you probably want to try a large array.

**Comment by 7dy109** [comment]
1. When we make Box<Vec<T>>, the pointer to data of vector, the data of vec is in heap, and the pointer to the pointer to data of vector is in stack right?
2. When we make Vec<T> data of vec is in already heap and pointer to the vec is in stack?
3. Then, if we move ownership of some Vec<T>, the data of vec is copied or not copied? 
4. And also, if we move ownership of some Box<Vec<T>>, the data of vec and the pointer of vector is copied or not copied? 
5. I did a bench with an array as you said. But the result is not equal I intended... What is the problem...?
```rust
    use std::time::Instant;
    fn takes_ownership(some : Box<[f32; 70000]>) {
        println!("Using Box<T>");
    }
    fn takes_ownership2(some : [f32;  70000]) {
        println!("Not Using Box<T>");
    }
    let start = Instant::now();
    let test:Box<[f32; 70000]> = Box::new([0.0; 70000]);
    takes_ownership(test);
    println!("{}", start.elapsed().as_nanos());
    let start2 = Instant::now();
    let test2 = [0.0; 70000];
    takes_ownership2(test2);
    println!("{}", start2.elapsed().as_nanos());
```
Result 
Sometimes, the time above is more than the time below.
```
rust
Using Box<T>
520200
Not Using Box<T>
393900
```
Sometimes, the time above is less than the time below
```
rust
Using Box<T>
405600
Not Using Box<T>
590500
```
6. During running the above code, if the size of the array is more than a specific value(I don't know the exact value, but it is about 70K~80K), there is panic with stack overflow.. What is the reason for this panic?
7. Thank you for your reply late midnight :)

**Comment by Lee-Janggun** [comment]
1. Yes
2. Yes
3. Not copied
4. Not copied
5. I'm not sure... maybe array is optimized enough so that there really isn't that much of an overhead.
6. Your are literally running out of memory.

**Comment by m-spitfire** [comment]
This comment mentions the use cases of `Box<T>`: https://users.rust-lang.org/t/when-should-i-use-box-t/4411/4

As the comment says, the case  you're investigating is very niche, so it might be hard to find an example. What I recommend you to do is, instead of passing it to a function, try to move by assigning variables, and instead of using time, try to generate assembly output, and analyze it. There if it's a correct case, for non-box you'll see `memcpy` for every move, and for box you might not. I recommend using https://godbolt.org/

**Comment by zzz845zz** [comment]
@7dy109 This example may show the advantage you expected.: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7f0d5b45a03c828e8beb0c611d7967c9

```
Using Box<T>
35381
Not Using Box<T>
115933
```

The difference with your example is: your example includes the overhead of dynamic heap allocation (i.e. initializing `Box`).

**Comment by 7dy109** [comment]
@Lee-Janggun  Thank you for your nice answer! I cleared my confusion about memory allocation of `Vec<_>` or `Box<Vec<_>>`! 
@m-spitfire Oops.. It was a niche... Because of that reason, It was hard to find such other examples... I can find another example by using godbolt by checking whether memcpy exist! I will try next time.(Assembly isn't familiar to me now.) Thank you for your good information.
@zzz845zz Oh, my god,,,, In my code, I included the declaring `test` or `test2` value to time measured... Thank you for your clarification.
I very much appreciate your kind answer. Thank you!!!
